polars/polars-core/src/chunked_array/kernels/mod.rs

```diff
@@ -7,12 +7,44 @@ pub(crate) mod take_agg;
 #[cfg_attr(docsrs, doc(cfg(feature = "temporal")))]
 pub mod temporal;
 
-use crate::datatypes::PolarsFloatType;
+use crate::datatypes::{DataType, PolarsFloatType, PolarsNumericType};
 use arrow::array::{Array, ArrayRef, BooleanArray, PrimitiveArray};
 use arrow::bitmap::Bitmap;
+use arrow::types::NativeType;
 use num::Float;
 use std::sync::Arc;
 
+/// Casts a `PrimitiveArray` from its logical type to a logical type compatible
+/// with its physical type.
+/// This operation is `O(1)`
+/// # Panics
+/// Panics if `data_type` is not supported by type `T`
+pub(crate) fn cast_logical<T: NativeType>(arr: &PrimitiveArray<T>, data_type: &DataType) -> ArrayRef
+where
+    T: NativeType,
+{
+    Arc::new(PrimitiveArray::<T>::from_data(
+        data_type.to_arrow(),
+        arr.values().clone(),
+        arr.validity().clone(),
+    ))
+}
+
+/// Casts a `PrimitiveArray` to a different physical type and logical type.
+/// This operation is `O(N)`
+/// Values that do not fit in the new physical type are converted to nulls.
+pub(crate) fn cast_physical<S, T>(arr: &PrimitiveArray<S>, datatype: &DataType) -> ArrayRef
```

Comment:
So an `i32` to `i8` will be a Null array?

Reply:
individual values. E.g. `&[Some(111111i32), None, Some(10)] -> &[None, None, Some(10i8)]`

Reply:
Of course. Check. :+1: